File: client/Core/Models/AppSession.cs
namespace client.Core.Models;

public class AppSession
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    public string ProcessName { get; set; } = string.Empty;
    public string AppDisplayName { get; set; } = string.Empty;
    public DateTime StartedAt { get; set; } = DateTime.UtcNow;
    public DateTime? EndedAt { get; set; }
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;

    public string MachineId { get; set; } = string.Empty;
    public string? EmployeeId { get; set; }
    public string? EmployeeName { get; set; }
    public string SessionId { get; set; } = string.Empty;
    public string Platform { get; set; } = string.Empty;

    /// <summary>FK → installed_applications.id, set when this session maps to a known GUI app</summary>
    public string? InstalledAppId { get; set; }
    /// <summary>FK → installed_packages.id, set when this session maps to a known CLI package</summary>
    public string? InstalledPackageId { get; set; }

    /// <summary>OS process ID at session start (for hierarchy linking across cycles)</summary>
    public int? ProcessId { get; set; }

    /// <summary>Parent OS process ID when nested under another session</summary>
    public int? ParentProcessId { get; set; }

    /// <summary>
    /// How this session's identity was grouped: 'cgroup' (systemd scope — the preferred
    /// path), 'pid' (fallback when the process has no app-*.scope), or null for pre-fix rows.
    /// </summary>
    public string? GroupedBy { get; set; }

    /// <summary>
    /// Raw systemd transient scope (e.g. "app-gnome-code-3a2f91.scope") that this session
    /// was grouped under. Needed at boot hydration to rebuild the scope-aware session key
    /// for still-running processes without duplicating sessions.
    /// </summary>
    public string? CgroupScope { get; set; }

    /// <summary>
    /// Human label to disambiguate multiple windows/instances of the same app:
    /// VS Code workspace/project folder name, Chrome --profile-directory, etc.
    /// </summary>
    public string? ContextLabel { get; set; }

    /// <summary>
    /// Accumulated seconds this session's window held the OS foreground focus.
    /// Null = not tracked (pre-2026-08-15 rows / non-GUI sessions). Grows while the
    /// session is open and is flushed periodically + on close (re-synced to the server).
    /// </summary>
    public double? ForegroundSeconds { get; set; }

    /// <summary>
    /// Accumulated seconds this session ran while another window had the focus.
    /// </summary>
    public double? BackgroundSeconds { get; set; }
}

/// <summary>
/// Generic child item for app_sessions. Replaces browser_contexts, file_explorer_contexts,
/// urls, and url_visits with a single self-referencing relational table.
/// </summary>
public class AppItem
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    /// <summary>FK → app_sessions.id</summary>
    public string AppSessionId { get; set; } = string.Empty;
    /// <summary>FK → app_items.id (nullable, for nesting: tab → terminal, tab → navigation, folder → subfolder)</summary>
    public string? ParentItemId { get; set; }
    /// <summary>Type discriminator: 'tab', 'browser_tab', 'browser_navigation', 'terminal', 'folder', 'file'</summary>
    public string ItemType { get; set; } = string.Empty;
    /// <summary>Display title: page title, file name, tab name, folder name, etc.</summary>
    public string Title { get; set; } = string.Empty;
    /// <summary>Content identifier: URL, file path, folder path, shell command, etc.</summary>
    public string Identifier { get; set; } = string.Empty;
    /// <summary>Full URL for browser_navigation items (from browser extension)</summary>
    public string Url { get; set; } = string.Empty;
    /// <summary>Extracted domain from URL for browser_navigation items</summary>
    public string Domain { get; set; } = string.Empty;
    public DateTime OpenedAt { get; set; } = DateTime.UtcNow;
    public DateTime? ClosedAt { get; set; }
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;

    /// <summary>OS process ID for process/runtime items nested under a terminal</summary>
    public int? ProcessId { get; set; }

    // ── Journey Event Fields (coexist with item_type) ──

    /// <summary>Generic object type: "Folder", "File", "Window", "Tab", "Page", etc.</summary>
    public string ObjectType { get; set; } = string.Empty;
    /// <summary>Action performed: "navigate", "open", "close", "create", "delete", "rename", etc.</summary>
    public string Action { get; set; } = string.Empty;
    /// <summary>Groups events into a journey (one per window/tab session)</summary>
    public string JourneyId { get; set; } = string.Empty;
    /// <summary>Deterministic ordering within a journey</summary>
    public int Sequence { get; set; }
    /// <summary>Previous location (for navigation reconstruction)</summary>
    public string PreviousPath { get; set; } = string.Empty;
    /// <summary>Current location</summary>
    public string CurrentPath { get; set; } = string.Empty;
    /// <summary>OS window identifier</summary>
    public int? WindowId { get; set; }
    /// <summary>Internal tab identifier within a window</summary>
    public int? TabId { get; set; }
    /// <summary>Flexible JSON blob for extensible metadata</summary>
    public string MetadataJson { get; set; } = "{}";
}

File: server/internal/models/app_session.go
package models

import "time"

type AppSession struct {
	ID               string     `json:"id" db:"id"`
	EmployeeID       string     `json:"employeeId" db:"employee_id"`
	ProcessName      string     `json:"processName" db:"process_name"`
	AppDisplayName   string     `json:"appDisplayName" db:"app_display_name"`
	StartedAt        time.Time  `json:"startedAt" db:"started_at"`
	EndedAt          *time.Time `json:"endedAt,omitempty" db:"ended_at"`
	MachineID        string     `json:"machineId" db:"machine_id"`
	SessionID        string     `json:"sessionId" db:"session_id"`
	Platform         string     `json:"platform" db:"platform"`
	ProcessID        *int       `json:"processId,omitempty" db:"process_id"`
	ParentProcessID  *int       `json:"parentProcessId,omitempty" db:"parent_process_id"`
	InstalledAppID   *string    `json:"installedAppId,omitempty" db:"installed_app_id"`
	InstalledPackageID *string  `json:"installedPackageId,omitempty" db:"installed_package_id"`
	GroupedBy        *string    `json:"groupedBy,omitempty" db:"grouped_by"`
	CgroupScope      *string    `json:"cgroupScope,omitempty" db:"cgroup_scope"`
	ContextLabel     *string    `json:"contextLabel,omitempty" db:"context_label"`
	SyncedAt         *time.Time `json:"syncedAt,omitempty" db:"synced_at"`
	CreatedAt        time.Time  `json:"createdAt" db:"created_at"`
}

// AppItem is a generic child of AppSession, replacing BrowserContext, FileExplorerContext, UrlRecord, UrlVisit.
// Self-referencing via ParentItemID for nesting: app_session -> tab -> terminal/browser_navigation
// ItemType: 'tab', 'browser_tab', 'browser_navigation', 'terminal', 'folder', 'file', etc.
type AppItem struct {
	ID           string     `json:"id" db:"id"`
	EmployeeID   string     `json:"employeeId" db:"employee_id"`
	AppSessionID string     `json:"appSessionId" db:"app_session_id"`
	ParentItemID *string    `json:"parentItemId,omitempty" db:"parent_item_id"`
	ItemType     string     `json:"itemType" db:"item_type"`
	Title        string     `json:"title" db:"title"`
	Identifier   string     `json:"identifier" db:"identifier"`
	Url          string     `json:"url" db:"url"`
	Domain       string     `json:"domain" db:"domain"`
	OpenedAt     time.Time  `json:"openedAt" db:"opened_at"`
	ClosedAt     *time.Time `json:"closedAt,omitempty" db:"closed_at"`
	ProcessID    *int       `json:"processId,omitempty" db:"process_id"`
	ObjectType   string     `json:"objectType" db:"object_type"`
	Action       string     `json:"action" db:"action"`
	JourneyID    string     `json:"journeyId" db:"journey_id"`
	Sequence     int        `json:"sequence" db:"sequence"`
	PreviousPath string     `json:"previousPath" db:"previous_path"`
	CurrentPath  string     `json:"currentPath" db:"current_path"`
	WindowID     *int       `json:"windowId,omitempty" db:"window_id"`
	TabID        *int       `json:"tabId,omitempty" db:"tab_id"`
	MetadataJSON string     `json:"metadataJson" db:"metadata_json"`
	SyncedAt     *time.Time `json:"syncedAt,omitempty" db:"synced_at"`
	CreatedAt    time.Time  `json:"createdAt" db:"created_at"`
}
